nixos/modules/services/networking/pppd.nix

```diff
@@ -0,0 +1,141 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.pppd;
+in
+{
+  meta = {
+    maintainers = with maintainers; [ danderson ];
+  };
+
+  options = {
+    services.pppd = {
+      enable = mkEnableOption "pppd";
+
+      package = mkOption {
+        default = pkgs.ppp;
+        defaultText = "pkgs.ppp";
+        type = types.package;
+        description = "pppd package to use.";
+      };
+
+      peers = mkOption {
+        default = {};
+        type = types.attrsOf (types.submodule (
+          { name, ... }:
+          {
+            options = {
+              name = mkOption {
+                type = types.str;
+                default = name;
+                example = "dialup";
+                description = "Name of the PPP peer.";
+              };
+
+              enable = mkOption {
+                type = types.bool;
+                default = true;
+                example = false;
+                description = "Whether to enable this PPP peer.";
+              };
+
+              autostart = mkOption {
+                type = types.bool;
+                default = true;
+                example = false;
+                description = "Whether the PPP session is automatically started at boot time.";
+              };
+
+              config = mkOption {
+                type = types.lines;
+                default = "";
+                description = "pppd configuration for this peer, see the pppd(8) man page.";
+              };
+            };
+          }));
+      };
+    };
+  };
+
+  config = let
+    # pppd likes to write directly into /var/run. This is rude on a
+    # modern system, so this wrapper transparently moves those files
+    # into /run/pppd.
+    pppd-wrapped = pkgs.stdenv.mkDerivation {
+      name = "pppd-wrapped";
+      phases = [ "installPhase" ];
+      buildInputs = [ pkgs.makeWrapper ];
+      installPhase = ''
+        mkdir -p $out/sbin
+        makeWrapper ${getBin cfg.package}/sbin/pppd $out/sbin/pppd \
+          --set LD_PRELOAD "${pkgs.libredirect}/lib/libredirect.so" \
+          --set NIX_REDIRECTS "/var/run=/run/pppd"
+      '';
+    };
+
+    enabledConfigs = filter (f: f.enable) (attrValues cfg.peers);
+
+    mkEtc = peerCfg: {
+      "ppp/peers/${peerCfg.name}".text = peerCfg.config;
+    };
+
+    mkSystemd = peerCfg: {
+      "pppd-${peerCfg.name}" = {
+        restartTriggers = [ config.environment.etc."ppp/peers/${peerCfg.name}".source ];
+        before = [ "network.target" ];
+        wants = [ "network.target" ];
+        after = [ "network-pre.target" ];
+        serviceConfig = {
+          ExecStart = "${getBin pppd-wrapped}/sbin/pppd call ${peerCfg.name} nodetach nolog";
```

Comment:
`pppd-wrapped` can be removed, by setting `"${peerCfg.name}".environment` - we don't use pppd-wrapped outside the systemd unit anyway.

Reply:
Good call! Initially I wasn't sure if I wanted pppd in systemPackages, but we decided we only care about systemd units, so this is much cleaner.

Done. Nixos test still passes, and my home internet still works.